src/java.base/share/classes/java/security/Signature.java

```diff
@@ -1005,17 +1005,18 @@ public final void setParameter(AlgorithmParameterSpec params)
     /**
      * Returns the parameters used with this signature object.
      *
-     * <p> If this signature has been initialized with parameters
-     * (by calling {@link #setParameter(AlgorithmParameterSpec)} or
-     * {@link #setParameter(String, Object)}) and the underlying signature
-     * implementation supports returning the parameters as
-     * {@code AlgorithmParameters}, this method returns the same parameters.
-     * If the parameters were not set, this method may return a combination
-     * of default and randomly generated parameter values if the
-     * underlying signature implementation supports it and can successfully
-     * generate them. Otherwise, {@code null} is returned.
+     * <p>The returned parameters may be the same that were used to initialize
+     * this signature, or may contain additional default or random parameter
+     * values used by the underlying signature implementation if the underlying
+     * signature implementation supports returning the parameters as
+     * {@code AlgorithmParameters}. If the required
+     * parameters were not supplied and the underlying signature implementation
+     * can generate the parameter values, it will be returned. Otherwise,
```

Comment:
> What does it refer to with 'it'? Is 'it' refer to the implementation generated parameter values?

It refers to the parameters containing all of the parameter values including the supplied ones and provider-generated ones if any.

Reply:
Can you clarify what is the A and B that you are referring to? The way I read it, it has more than 2 conditions... So, best to clarify the conditions first.

Reply:
With Signature class, there is a caveat for EdDSA, the supplied parameters are set but null is being returned when getParameters() is called. This is currently covered by the condition `if the underlying signature implementation supports returning the parameters as {@code AlgorithmParameters}` as the underlying signature does not support AlgorithmParameters for the supplied EdDSAParameterSpec object due to lack of ASN.1 definition.

Reply:
Besides this Signature-specific condition, there is the common condition where provider cannot (or do not) generate default parameter values. {@code null} is used as the catch-all result, but as you said, describe various conditions tersely and correctly is key.

Reply:
> > What does it refer to with 'it'? Is 'it' refer to the implementation generated parameter values?
> 
> 'It' refers to the parameters containing all of the parameter values including the supplied ones and provider-generated ones if any.

The full sentence is, "If the required parameters were not supplied and the underlying signature implementation can generate the parameter values, it will be returned."   As there is no supplied value, I think 'it' refer to the provider-generated ones if any.  As the previous noun is "the parameters values", I'm not sure if the use of 'it' here is properly.

Reply:
> Can you clarify what is the A and B that you are referring to?
The sentence is, “If the required parameters were not supplied and the underlying signature implementation can generate the parameter values, it will be returned. Otherwise, {https://github.com/code null} is returned."

I read "the required parameters were not supplied" as condition A; and "the underlying signature implementation can generate the parameter values" as condition B.

Reply:
> With Signature class, there is a caveat for EdDSA, the supplied parameters are set but null is being returned when getParameters() is called. This is currently covered by the condition `if the underlying signature implementation supports returning the parameters as {@code AlgorithmParameters}` as the underlying signature does not support AlgorithmParameters for the supplied EdDSAParameterSpec object due to lack of ASN.1 definition.

I see now.  My 1st read of the condition, I did not get the point and thought it is not necessary as the method is trying to return "AlgorithmParameters".  Could it be more clear if describe this case in an additional sentence?

Reply:
What kind of additional sentence do you have in mind?

Reply:
How about the case when no parameters are given? Say A is the user-supplied values, B is the provider specific default or random values, your suggestion has A, A+B, and null. Isn't the sentence about B needed (no A and provider can generate the parameters)?

Reply:
I think this sentence covers case B, "... or may contain additional default or random parameter
values used by the underlying signature implementation."

Reply:
Sean's comment on the other PR regarding Cipher.getParameters implies otherwise, i.e. `A few comments:

    I don't think this new text covers the case where the cipher is not initialized with any parameters, but the provider returns parameters. I tried to think of how to say that all in one sentence, but I think it is best to have two sentences covering each case, ex:

"If this cipher was not initialized with parameters, the returned parameters may be null or may be ..."
"If this cipher was initialized with parameters, the returned parameters may be the same or may be ..."`

Reply:
He later agrees that two sentences are probably too lengthy, so we ended up exploring other wording choices.